File: app/src/main/java/uk/co/appsbystudio/connect/MainActivity.java
package uk.co.appsbystudio.connect;

import android.app.Activity;
import android.os.Bundle;
import android.renderscript.Sampler;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import uk.co.appsbystudio.connect.server.ClientThread;
import uk.co.appsbystudio.connect.server.ServerConnect;

public class MainActivity extends Activity {

    ClientThread clientThread;
    Thread thread;
    ServerConnect serverConnect;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        final EditText editTextAddress = (EditText) findViewById(R.id.editTextAddress);
        final EditText editTextPort = (EditText) findViewById(R.id.editTextPort);

        findViewById(R.id.connectButton).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //serverConnect = new ServerConnect(editTextAddress.getText().toString(), Integer.valueOf(String.valueOf(editTextPort.getText())));
                //serverConnect.execute();
                if (thread == null || !thread.isAlive()) {
                    clientThread = new ClientThread(editTextAddress.getText().toString(), Integer.valueOf(String.valueOf(editTextPort.getText())));
                    thread = new Thread(clientThread);
                    thread.start();
                } else System.out.println("Already connected!");
            }
        });

        findViewById(R.id.messageSend).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Thread thread = new Thread(new Runnable() {
                    @Override
                    public void run() {
                        clientThread.sendMessage("This should work!");
                    }
                });
                thread.start();
            }
        });

        findViewById(R.id.disconnectButton).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                clientThread.closeConnection();
            }
        });
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (thread != null) {
            clientThread.closeConnection();
        }
    }
}

File: app/src/main/java/uk/co/appsbystudio/connect/MainActivity.kt
package uk.co.appsbystudio.connect

import androidx.lifecycle.Observer
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import com.google.android.material.bottomnavigation.BottomNavigationView
import androidx.appcompat.app.AppCompatActivity
import android.view.MenuItem
import android.view.View
import android.view.animation.AnimationUtils
import android.widget.Toast
import androidx.navigation.Navigation
import androidx.navigation.ui.setupWithNavController
import kotlinx.android.synthetic.main.activity_main.*
import uk.co.appsbystudio.connect.data.AppDatabase
import uk.co.appsbystudio.connect.data.models.ServerModel
import uk.co.appsbystudio.connect.ui.dashboard.DashboardFragment
import uk.co.appsbystudio.connect.ui.server.ServerFragment
import uk.co.appsbystudio.connect.ui.settings.SettingsFragment
import uk.co.appsbystudio.connect.utils.OnSocketStateChangeListener
import uk.co.appsbystudio.connect.utils.SocketService
import uk.co.appsbystudio.connect.utils.SocketUtils

class MainActivity : AppCompatActivity(), OnSocketStateChangeListener.SocketStateReceiverListener {

    private var connected = false

    private var socketUtils: SocketUtils? = null
    private var onSocketStateChangeListener = OnSocketStateChangeListener()

    lateinit var service: Intent

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val appDatabase = AppDatabase.getInstance(this.application)

        setSupportActionBar(toolbar)

        socketUtils = SocketUtils()

        service = Intent(this, SocketService::class.java)

        startService(service)

        appDatabase.serverDao().observerSelected().observe(this, Observer<ServerModel>() {
            stopService(service)
            startService(service)
            println(it?.uid)
        })

        val navController = Navigation.findNavController(this, R.id.nav_host_fragment)

        bottom_navigation_main.setupWithNavController(navController)

        registerReceiver(onSocketStateChangeListener, IntentFilter("socket.state"))
        onSocketStateChangeListener.addListener(this)
    }

    override fun onDestroy() {
        super.onDestroy()

        socketUtils?.closeSocket(this)

        unregisterReceiver(onSocketStateChangeListener)
        onSocketStateChangeListener.removeListener(this)
    }

    override fun socketConnected() {
        connected = true
        Toast.makeText(this, "Connected to server", Toast.LENGTH_SHORT).show()

    }

    override fun socketDisconnected() {
        connected = false
        Toast.makeText(this, "Disconnected from server", Toast.LENGTH_SHORT).show()
    }
}
